File: topic_example/test/test_publisher.py
# # Copyright 2019 Open Source Robotics Foundation, Inc.
# #
# # Licensed under the Apache License, Version 2.0 (the "License");
# # you may not use this file except in compliance with the License.
# # You may obtain a copy of the License at
# #
# #     http://www.apache.org/licenses/LICENSE-2.0
# #
# # Unless required by applicable law or agreed to in writing, software
# # distributed under the License is distributed on an "AS IS" BASIS,
# # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# # See the License for the specific language governing permissions and
# # limitations under the License.
# #
# # Authors: Ryan Shim

# import os

# import unittest

# from launch import LaunchDescription
# from launch.actions import ExecuteProcess

# import launch_testing
# import launch_testing.actions
# import launch_testing.asserts
# import launch_testing.util
# import launch_testing_ros


# def generate_test_description():
#     env = os.environ.copy()
#     env['OSPL_VERBOSITY'] = '8'  # 8 = OS_NONE
#     # bare minimum formatting for console output matching
#     env['RCUTILS_CONSOLE_OUTPUT_FORMAT'] = '{message}'

#     launch_description = LaunchDescription()
#     processes_under_test = [
#         ExecuteProcess(
#             cmd=[executable],
#             name='test_executable_' + str(i),
#             output='screen',
#             env=env)
#         for i, executable in enumerate('@DEMO_NODES_CPP_EXECUTABLE@'.split(';'))
#     ]
#     for process in processes_under_test:
#         launch_description.add_action(process)
#     launch_description.add_action(launch_testing.util.KeepAliveProc())
#     launch_description.add_action(
#         launch_testing.actions.ReadyToTest()
#     )
#     return launch_description, locals()


# class TestExecutablesTutorial(unittest.TestCase):

#     def test_processes_output(self, proc_output, processes_under_test):
#         """Test all processes output against expectations."""
#         from launch_testing.tools.output import get_default_filtered_prefixes
#         output_filter = launch_testing_ros.tools.basic_output_filter(
#             filtered_prefixes=get_default_filtered_prefixes() + [
#                 'service not available, waiting again...'
#             ],
#             filtered_rmw_implementation='@rmw_implementation@'
#         )
#         output_files = '@DEMO_NODES_CPP_EXPECTED_OUTPUT@'.split(';')
#         for process, output_file in zip(processes_under_test, output_files):
#             proc_output.assertWaitFor(
#                 expected_output=launch_testing.tools.expected_output_from_file(
#                     path=output_file
#                 ), process=process, output_filter=output_filter, timeout=30
#             )
#         # TODO(hidmic): either make the underlying executables resilient to
#         # interruptions close/during shutdown OR adapt the testing suite to
#         # better cope with it.
#         import time
#         time.sleep(5)


# @launch_testing.post_shutdown_test()
# class TestExecutablesTutorialAfterShutdown(unittest.TestCase):

#     def test_last_process_exit_code(self, proc_info, processes_under_test):
#         """Test last process exit code."""
#         launch_testing.asserts.assertExitCodes(
#             proc_info,
#             process=processes_under_test[-1]
#         )

File: topic_example/test/test_publisher.cpp
// Copyright 2020 ROBOTIS CO., LTD.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/* Authors: Ryan Shim */

#include <gtest/gtest.h>

#include <chrono>
#include <memory>
#include <utility>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_components/register_node_macro.hpp"

#include "interface_examples/msg/count.hpp"
#include "topic_example/visibility_control.h"

using namespace std::chrono_literals;

namespace topic_example
{
class Publisher : public rclcpp::Node
{
public:
  Publisher()  // options for what ???
  : Node("publisher")
  {
    // Force flush of the stdout buffer
    setvbuf(stdout, NULL, _IONBF, BUFSIZ);

    // ROS Publisher
    rclcpp::QoS qos(rclcpp::KeepLast(10));
    count_pub_ = this->create_publisher<interface_examples::msg::Count>("chatter", qos);

    // ROS Timer
    auto timer_callback =
      [this]() -> void
      {
        msg_ = std::make_unique<interface_examples::msg::Count>();
        msg_->data = count_++;
        RCLCPP_INFO(this->get_logger(), "%d", msg_->data);
        count_pub_->publish(std::move(msg_));
      };
    timer_ = this->create_wall_timer(1s, timer_callback);

    RCLCPP_INFO(this->get_logger(), "Initialized publisher node");
  }
  ~Publisher()
  {
    RCLCPP_INFO(this->get_logger(), "Terminated publisher node");
  }

private:
  uint16_t count_ = 0;
  std::unique_ptr<interface_examples::msg::Count> msg_;  // why unique pointer for this?
  rclcpp::Publisher<interface_examples::msg::Count>::SharedPtr count_pub_;
  rclcpp::TimerBase::SharedPtr timer_;
};
}  // namespace topic_example

TEST(TestPublisher, test_publisher)
{
//   // rclcpp::init(0, nullptr);

//   // TODO(Ryan): Study how to write more general gtest codes!!!
//   // Also below hasn't been debugged yet
//   // <<< error message
//   // The test did not generate a result file.
//   // >>>

  EXPECT_EQ(1, 1);

//   // EXPECT_NO_THROW(
//   //   auto node = std::make_shared<topic_example::Publisher>();
//   // );
}

TEST(TestSuite, initTest)
{
  ASSERT_TRUE(true);
}

int main(int argc, char ** argv)
{
  testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}
